### ISSUE TITLE
Misc findings from the 7.0 beta

### ISSUE THREAD

**Issue by jdelStrother** [mperham/sidekiq#5579]
I found a grab-bag of little issues from the 7.0 beta. Most are probably user-error or issues with third-party gems, and happy for them to be ignored. But in case you want some feedback on how 7.0 is going, thought I'd mention them. Let me know if you want any broken out into a separate issue with more detail....

Ruby version: 3.1.2
Rails version: 7.0.3
Sidekiq / Pro / Enterprise version(s): 
sidekiq (= 7.0.0.beta1)
sidekiq-ent (= 3.0.0.beta1)
sidekiq-pro (= 6.0.0.beta1)

---

# Documentation 404

On installing the gem it tells me to check the release notes at https://github.com/mperham/sidekiq/blob/7-0/docs/7.0-Upgrade.md, which currently 404s.

---

# sentry-sidekiq gem

[sentry-sidekiq](https://github.com/getsentry/sentry-ruby) doesn't seem compatible with 7.0 : 
```
2022-10-20T16:35:14.279Z pid=47848 tid=17x8 ERROR: !!! ERROR HANDLER THREW AN ERROR !!!
2022-10-20T16:35:14.279Z pid=47848 tid=17x8 ERROR: wrong number of arguments (given 3, expected 2)
2022-10-20T16:35:14.279Z pid=47848 tid=17x8 ERROR: GEM_HOME/sentry-sidekiq-5.5.0/lib/sentry/sidekiq/error_handler.rb:6:in `call'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/config.rb:261:in `block in handle_exception'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/config.rb:260:in `each'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/config.rb:260:in `handle_exception'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:46:in `handle_exception'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:60:in `rescue in block in fire_event'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:58:in `block in fire_event'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:57:in `each'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:57:in `fire_event'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/senate.rb:85:in `stage_coup!'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/senate.rb:93:in `cycle'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:8:in `watchdog'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:17:in `block in safe_thread'
```

---

# Redis options

Looks like redis-client accepts slightly different initialization options for sentinels. redis-rb allows setting the cluster name in the `url` option, like so - 

```ruby
Sidekiq.configure_server do |config|
  config.redis = {:url=>"redis://vanilla", :driver=>:hiredis, :sentinels=>[{:host=>"127.0.0.1", :port=>"26379"}]}
```

whereas redis-client needed updating to - 
```ruby
Sidekiq.configure_server do |config|
  config.redis = {:name=>"vanilla", :sentinels=>[{:host=>"127.0.0.1", :port=>"26379"}]}
```

(definitely not sidekiq's fault, but maybe worth pointing out that the redis configuration might be different in the upgrade notes?)

---

# Periodic args

Arguments in periodic jobs seem broken. We have periodic jobs set up like this: 

```ruby
Sidekiq.configure_server do |config|
  config.periodic do |mgr|
    mgr.register("0 12 * * *", "FeedsPollerWorker", queue: "low", args: ["priority" => -1])
  end
end
```

which results in 

```
2022-10-20T16:43:41.978Z pid=51626 tid=113y WARN: {"context":"Exception during Sidekiq lifecycle event.","event":"leader"}
2022-10-20T16:43:41.978Z pid=51626 tid=113y WARN: TypeError: Unsupported command argument type: Array
2022-10-20T16:43:41.978Z pid=51626 tid=113y WARN:
GEM_HOME/redis-client-0.10.0/lib/redis_client/command_builder.rb:37:in `block in generate'
GEM_HOME/redis-client-0.10.0/lib/redis_client/command_builder.rb:28:in `map!'
GEM_HOME/redis-client-0.10.0/lib/redis_client/command_builder.rb:28:in `generate'
GEM_HOME/redis-client-0.10.0/lib/redis_client.rb:451:in `call'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/redis_client_adapter.rb:25:in `method_missing'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/periodic/config.rb:80:in `block (3 levels) in persist'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/periodic/config.rb:70:in `each'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/periodic/config.rb:70:in `block (2 levels) in persist'
GEM_HOME/redis-client-0.10.0/lib/redis_client/decorator.rb:51:in `block in pipelined'
GEM_HOME/redis-client-0.10.0/lib/redis_client.rb:344:in `pipelined'
GEM_HOME/redis-client-0.10.0/lib/redis_client/decorator.rb:51:in `pipelined'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/periodic/config.rb:66:in `block in persist'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/config.rb:156:in `block in redis'
GEM_HOME/connection_pool-2.3.0/lib/connection_pool.rb:65:in `block (2 levels) in with'
GEM_HOME/connection_pool-2.3.0/lib/connection_pool.rb:64:in `handle_interrupt'
GEM_HOME/connection_pool-2.3.0/lib/connection_pool.rb:64:in `block in with'
GEM_HOME/connection_pool-2.3.0/lib/connection_pool.rb:61:in `handle_interrupt'
GEM_HOME/connection_pool-2.3.0/lib/connection_pool.rb:61:in `with'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/config.rb:153:in `redis'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:26:in `redis'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/periodic/config.rb:65:in `persist'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/periodic/manager.rb:61:in `persist'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/periodic.rb:96:in `block (2 levels) in <main>'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:58:in `block in fire_event'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:57:in `each'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:57:in `fire_event'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/senate.rb:85:in `stage_coup!'
GEM_HOME/sidekiq-ent-3.0.0.beta1/lib/sidekiq-ent/senate.rb:93:in `cycle'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:8:in `watchdog'
GEM_HOME/sidekiq-7.0.0.beta1/lib/sidekiq/component.rb:17:in `block in safe_thread'
```

stepping back up to sidekiq's hmset call at sidekiq-ent/periodic/config.rb:80, `args` is: 

```ruby
["loop-0366e18552efecb2ca370e70f93c31b8601a0556", "schedule", "0 12 * * *", "class", "ChannelFeedsPollerWorker", ["queue", "low"], ["args", [{"priority"=>-1}]]]
```

are we misconfiguring those periodic job arguments somehow?

---

That aside, our test-suite is green and it seems ok in development. Not tried production yet....

**Comment by mperham** [comment]
Thanks so much for this feedback. I can't do much about redis-client but I bet @casperisfine can.

**Comment by mperham** [comment]
All issues fixed except for redis-client sentinel concern.

**Comment by casperisfine** [comment]
Unless I misunderstood and the TypeError was fixed in https://github.com/mperham/sidekiq/commit/02153c17360e712d9a94c08406fe7c057c4d7635 ?

**Comment by mperham** [comment]
```diff
-              args.append(*lop.options.to_a)
+              args.append("options", Sidekiq.dump_json(lop.options))
               p.hmset(*args)
```

**Comment by bf4** [comment]
I know this is closed, but seems related and small enough to merit a comment, a difference gem version constraints allow sidekiq-pro 5.5.4 to be compatible with 7.0, but not 5.5.5. Not sure if that's intentional.

> Updating files in vendor/cache
>   * redis-client-0.10.0.gem
>   * sidekiq-7.0.0.beta1.gem
>   * sidekiq-pro-5.5.4.gem
> Removing outdated .gem files from vendor/cache
>   * sidekiq-6.5.7.gem
>   * sidekiq-pro-5.5.5.gem
> Note: sidekiq-pro version regressed from 5.5.5 to 5.5.4

```diff
-    sidekiq-pro (5.5.5)
-      sidekiq (~> 6.0, >= 6.5.6)
+    sidekiq-pro (5.5.4)
+      sidekiq (>= 6.5.6)

-    sidekiq (6.5.7)
-      connection_pool (>= 2.2.5)
-      rack (~> 2.0)
-      redis (>= 4.5.0, < 5)
+    sidekiq (7.0.0.beta1)
+      concurrent-ruby (< 2)
+      connection_pool (>= 2.3.0)
+      rack (>= 2.2.4)
+      redis-client (>= 0.9.0)

+  sidekiq (= 7.0.0.beta1)
```

**Comment by jdelStrother** [comment]
> As for sentinel, yeah, the args changed a bit, I thought the name in URL was very confusing. If you think it was a good feature please make a case in an issue on redis-client and I will seriously consider it.

No, the cluster name as the URL host is pretty weird, happy to lose it.
(Though it's kind of a shame you can't specify the sentinel redis configuration from a single environment variable like you can for regular redis (`redis://1.2.3.4:6379/5`), but I'm not sure what that would look like for Sentinels - some sort of mess like `redis-sentinel://1.2.3.4:6379/1.2.3.5:6379/1.2.3.6:6379?db=5&name=master`)

**Comment by mperham** [comment]
> I know this is closed, but seems related and small enough to merit a comment, a difference gem version constraints allow sidekiq-pro 5.5.4 to be compatible with 7.0, but not 5.5.5. Not sure if that's intentional.
> 

Yes, there was a bug in 5.5.4 that was fixed in 5.5.5: it thinks it can run with 7.

**Comment by mperham** [comment]
@bf4 I can see this becoming a huge support issue so I've added a substantial compatibility check which should make the issue clear to the developer: https://github.com/mperham/sidekiq/commit/b4092e33f97e32c0e4e3f57c38c351867c5bc87b
